torch/_functorch/_aot_autograd/fx_passes.py

```diff
@@ -0,0 +1,152 @@
+from collections import defaultdict
+from typing import Dict, List
+
+import torch
+from torch import fx
+
+
+def is_primal(node: fx.Node) -> bool:
+    return node.op == "placeholder" and "tangents" not in str(node.target)
+
+
+def move_resize_zero_to_end_of_graph(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    for n in node_list:
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.inductor.resize_storage_bytes_.default
+            and n.args[1] == 0
+        ):
+            resize_node = n
+            with graph.inserting_before(return_node):
+                new_resize_node = graph.call_function(
+                    torch.ops.inductor.resize_storage_bytes_.default, resize_node.args
+                )
+            graph.erase_node(resize_node)
+
+
+def refunctionalize_set(graph: fx.Graph) -> None:
+    node_list = list(graph.nodes)
+    return_node = node_list[-1]
+    assert return_node.target == "output"
+    primal_inputs = [*filter(is_primal, node_list)]
+    primal_input_to_set_idx = defaultdict(list)
+    node_to_idx = {n: i for i, n in enumerate(node_list)}
+    set_nodes_to_be_deleted = set()
+
+    # Step 1: Clean up to enforce `.set_(primal_X, ...)` usage.
+    # We replace:
+    # ```
+    # set_1 = aten.set_(primal_X, Y1)
+    # set_2 = aten.set_(set_1, Y2)
+    # ```
+    # to:
+    # ```
+    # set_1 = aten.set_(primal_X, Y1)
+    # set_2 = aten.set_(primal_X, Y2)
+    # ```
+    for i, n in enumerate(node_list):
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.aten.set_.source_Tensor
+            and n.args[0].target is torch.ops.aten.set_.source_Tensor
+            and n.args[0].args[0] in primal_inputs
+        ):
+            n.args = (n.args[0].args[0], n.args[1])
```

Comment:
Are we only supporting two "levels" of set_() here because that's all that FSDP needs?

I'm hoping that we get a reasonable error whenever we end up with a graph that these graph transforms didn't expect. So in this for loop, should we add some extra asserts like "every `set_` node that we encounter always has an input that is a primal"? (I think that is the invariant that this loop is trying to maintain?)

Reply:
We should add strong checks on this, because `.set_` in user code graph-break now (need to check)

Reply:
> every set_ node that we encounter always has an input that is a primal

added an assertion to check this